README.md
# Dashed Rect

This package allows you to draw dotted lines with Flutter. You can draw a beautiful dotted line.

![Bitbucket open issues](https://img.shields.io/bitbucket/issues-raw/champ96k/DashedRect)

## Screenshots

![](https://i.ibb.co/kmYrfcM/screenshot.png)

## Usage

[Example](https://github.com/champ96k/DashedRect/tree/master/example)

To use this package :

- add the dependency to your [pubspec.yaml](https://github.com/champ96k/DashedRect/tree/master/example/pubspec.yaml) file.

```yaml
dependencies:
  flutter:
    sdk: flutter
  dashed_rect: ^0.0.4
```

### How to use

```dart
 DashedRect(
    gap: 8,
    strokeWidth: 8.0,
    color: Colors.white,
    child: yourWidget,
  ),
```

### Created & Maintained By

[Tushar Nikam](https://champ96k.github.io)

<a href="https://www.twitter.com/champ_96k"><img src="https://img.shields.io/badge/twitter-%231DA1F2.svg?&style=for-the-badge&logo=twitter&logoColor=white" height=25> </a>
<br>
<a href="https://www.linkedin.com/in/tushar-nikam-a29a97131/"><img src="https://img.shields.io/badge/linkedin-%230077B5.svg?&style=for-the-badge&logo=linkedin&logoColor=white" height=25></a>


CHANGELOG.md
## 0.0.1

A Flutter package to easily add dashed borders around widgets

## 0.0.2

Improve documentation

## 0.0.3

Follow Dart file conventions

## 0.0.4

Follow Dart file conventions and added child widget support


test/dashed_rect_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:dashed_rect/dashed_rect.dart';

void main() {
  test('adds one to input values', () {
    const dashedRect = DashedRect(
      color: Colors.red,
      gap: 1.0,
      child: SizedBox(height: 20, width: 20.0),
    );
    expect(dashedRect.color, Colors.green);
  });
}


lib/example.dart
import 'package:flutter/material.dart';

import 'dashed_rect.dart';

class ExampleApp extends StatelessWidget {
  const ExampleApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Dashed Rect example"),
      ),
      body: Center(
        child: SingleChildScrollView(
          child: Column(
            children: [
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: const [
                  DashedRect(
                    gap: 2,
                    strokeWidth: 1,
                    child: SizedBox(
                      height: 100,
                      width: 100,
                    ),
                  ),
                  DashedRect(
                    gap: 2,
                    strokeWidth: 1,
                    color: Colors.indigo,
                    child: SizedBox(
                      height: 100,
                      width: 100,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  DashedRect(
                    gap: 2,
                    strokeWidth: 1,
                    child: Container(
                      height: 100,
                      width: 100,
                      color: Colors.green,
                    ),
                  ),
                  DashedRect(
                    gap: 2,
                    strokeWidth: 6,
                    color: Colors.red,
                    child: Container(
                      height: 100,
                      width: 100,
                      color: Colors.indigo,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  DashedRect(
                    gap: 12,
                    strokeWidth: 1,
                    child: Container(
                      height: 100,
                      width: 100,
                      color: Colors.yellow,
                    ),
                  ),
                  const DashedRect(
                    gap: 2,
                    strokeWidth: 12,
                    color: Colors.red,
                    child: SizedBox(
                      height: 100,
                      width: 100,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: const [
                  DashedRect(
                    gap: 12,
                    strokeWidth: 12,
                    child: SizedBox(
                      height: 100,
                      width: 100,
                    ),
                  ),
                  DashedRect(
                    gap: 12,
                    strokeWidth: 1,
                    color: Colors.red,
                    child: SizedBox(
                      height: 100,
                      width: 100,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: const [
                  DashedRect(
                    gap: 10,
                    strokeWidth: 10,
                    color: Colors.orange,
                    child: SizedBox(
                      height: 100,
                      width: 100,
                    ),
                  ),
                  DashedRect(
                    child: SizedBox(
                      height: 100,
                      width: 100,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  DashedRect(
                    gap: 8,
                    strokeWidth: 8.0,
                    color: Colors.white,
                    child: Container(
                      height: 100,
                      width: 100,
                      decoration: BoxDecoration(
                        color: Colors.blue,
                        borderRadius: BorderRadius.circular(12.0),
                      ),
                    ),
                  ),
                  DashedRect(
                    gap: 8,
                    strokeWidth: 8.0,
                    color: Colors.orange,
                    child: Container(
                      height: 100,
                      width: 100,
                      decoration: BoxDecoration(
                        color: Colors.indigo,
                        borderRadius: BorderRadius.circular(12.0),
                      ),
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
            ],
          ),
        ),
      ),
    );
  }
}


lib/dashed_rect.dart
library dashed_rect;

import 'dart:math' as math;

import 'package:flutter/material.dart';

/// A Flutter package to easily add dashed borders around widgets

/// The [strokeWidth] and [gap] argument must not be null.
/// default `strokeWidth = 1.0` and  `gap = 5.0`
///
class DashedRect extends StatelessWidget {
  /// Creates a DashedRect which contain doted line
  ///
  final Color color;
  final Widget? child;

  /// The amount of space by which to inset the child.
  final double strokeWidth;
  final double gap;

  /// The [strokeWidth] and [gap] argument must not be less than zero.
  const DashedRect({
    this.color = Colors.black,
    this.strokeWidth = 1.0,
    this.child,
    this.gap = 5.0,
    Key? key,
  })  : assert(strokeWidth > 0 && gap > 0),
        super(key: key);

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: EdgeInsets.all(strokeWidth / 2),
      child: CustomPaint(
        painter: DashRectPainter(
          color: color,
          strokeWidth: strokeWidth,
          gap: gap,
        ),
        child: child,
      ),
    );
  }
}

/// A widget that create the doted line itself.
class DashRectPainter extends CustomPainter {
  double strokeWidth;
  Color color;
  double gap;

  DashRectPainter({
    required this.strokeWidth,
    required this.color,
    required this.gap,
  });

  @override
  void paint(Canvas canvas, Size size) {
    Paint dashedPaint = Paint()
      ..color = color
      ..strokeWidth = strokeWidth
      ..style = PaintingStyle.stroke;

    double x = size.width;
    double y = size.height;

    Path _topPath = getDashedPath(
      a: const math.Point(0, 0),
      b: math.Point(x, 0),
      gap: gap,
    );

    Path _rightPath = getDashedPath(
      a: math.Point(x, 0),
      b: math.Point(x, y),
      gap: gap,
    );

    Path _bottomPath = getDashedPath(
      a: math.Point(0, y),
      b: math.Point(x, y),
      gap: gap,
    );

    Path _leftPath = getDashedPath(
      a: const math.Point(0, 0),
      b: math.Point(0.001, y),
      gap: gap,
    );

    canvas.drawPath(_topPath, dashedPaint);
    canvas.drawPath(_rightPath, dashedPaint);
    canvas.drawPath(_bottomPath, dashedPaint);
    canvas.drawPath(_leftPath, dashedPaint);
  }

  Path getDashedPath({
    required math.Point<double> a,
    required math.Point<double> b,
    required double gap,
  }) {
    Size size = Size(b.x - a.x, b.y - a.y);
    Path path = Path();
    path.moveTo(a.x, a.y);
    bool shouldDraw = true;
    math.Point currentPoint = math.Point(a.x, a.y);

    num radians = math.atan(size.height / size.width);

    num dx = math.cos(radians) * gap < 0
        ? math.cos(radians) * gap * -1
        : math.cos(radians) * gap;

    num dy = math.sin(radians) * gap < 0
        ? math.sin(radians) * gap * -1
        : math.sin(radians) * gap;

    while (currentPoint.x <= b.x && currentPoint.y <= b.y) {
      shouldDraw
          ? path.lineTo(currentPoint.x as double, currentPoint.y as double)
          : path.moveTo(currentPoint.x as double, currentPoint.y as double);
      shouldDraw = !shouldDraw;
      currentPoint = math.Point(
        currentPoint.x + dx,
        currentPoint.y + dy,
      );
    }
    return path;
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }
}


example/example_app.dart
import 'package:dashed_rect/dashed_rect.dart';
import 'package:flutter/material.dart';

class ExampleApp extends StatelessWidget {
  const ExampleApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Column(
        children: [
          const SizedBox(
            height: 100,
            width: 100,
            child: DashedRect(
              gap: 1,
              strokeWidth: 1,
            ),
          ),
          const SizedBox(height: 16.0),
          Container(
            color: Colors.red,
            height: 600,
            width: 300,
            child: const DashedRect(
              color: Colors.grey,
              gap: 1,
              strokeWidth: 1,
            ),
          ),
          const SizedBox(height: 16.0),
        ],
      ),
    );
  }
}


lib/dashed_rect_example.dart
import 'package:flutter/material.dart';

import 'dashed_rect.dart';

class ExampleApp extends StatelessWidget {
  const ExampleApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Dashed Rect example"),
      ),
      body: Center(
        child: SingleChildScrollView(
          child: Column(
            children: [
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: const [
                  SizedBox(
                    height: 100,
                    width: 100,
                    child: DashedRect(
                      gap: 2,
                      strokeWidth: 1,
                    ),
                  ),
                  SizedBox(
                    height: 100,
                    width: 100,
                    child: DashedRect(
                      gap: 2,
                      strokeWidth: 1,
                      color: Colors.indigo,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  Container(
                    height: 100,
                    width: 100,
                    color: Colors.green,
                    child: const DashedRect(
                      gap: 2,
                      strokeWidth: 1,
                    ),
                  ),
                  Container(
                    height: 100,
                    width: 100,
                    color: Colors.indigo,
                    child: const DashedRect(
                      gap: 2,
                      strokeWidth: 6,
                      color: Colors.red,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  Container(
                    height: 100,
                    width: 100,
                    color: Colors.yellow,
                    child: const DashedRect(
                      gap: 12,
                      strokeWidth: 1,
                    ),
                  ),
                  const SizedBox(
                    height: 100,
                    width: 100,
                    child: DashedRect(
                      gap: 2,
                      strokeWidth: 12,
                      color: Colors.red,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: const [
                  SizedBox(
                    height: 100,
                    width: 100,
                    child: DashedRect(
                      gap: 12,
                      strokeWidth: 12,
                    ),
                  ),
                  SizedBox(
                    height: 100,
                    width: 100,
                    child: DashedRect(
                      gap: 12,
                      strokeWidth: 1,
                      color: Colors.red,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: const [
                  SizedBox(
                    height: 100,
                    width: 100,
                    child: DashedRect(
                      gap: 10,
                      strokeWidth: 10,
                      color: Colors.orange,
                    ),
                  ),
                  SizedBox(
                    height: 100,
                    width: 100,
                    child: DashedRect(),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  Container(
                    height: 100,
                    width: 100,
                    decoration: BoxDecoration(
                      color: Colors.blue,
                      borderRadius: BorderRadius.circular(12.0),
                    ),
                    child: const DashedRect(
                      gap: 8,
                      strokeWidth: 8.0,
                      color: Colors.white,
                    ),
                  ),
                  Container(
                    height: 100,
                    width: 100,
                    decoration: BoxDecoration(
                      color: Colors.indigo,
                      borderRadius: BorderRadius.circular(12.0),
                    ),
                    child: const DashedRect(
                      gap: 8,
                      strokeWidth: 8.0,
                      color: Colors.orange,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
            ],
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:dashed_rect/dashed_rect.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Dashed Rect example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: const ExampleApp(),
    );
  }
}

class ExampleApp extends StatelessWidget {
  const ExampleApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Dashed Rect example"),
      ),
      body: Center(
        child: SingleChildScrollView(
          child: Column(
            children: [
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: const [
                  SizedBox(
                    height: 100,
                    width: 100,
                    child: DashedRect(
                      gap: 2,
                      strokeWidth: 1,
                    ),
                  ),
                  SizedBox(
                    height: 100,
                    width: 100,
                    child: DashedRect(
                      gap: 2,
                      strokeWidth: 1,
                      color: Colors.indigo,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  Container(
                    height: 100,
                    width: 100,
                    color: Colors.green,
                    child: const DashedRect(
                      gap: 2,
                      strokeWidth: 1,
                    ),
                  ),
                  Container(
                    height: 100,
                    width: 100,
                    color: Colors.indigo,
                    child: const DashedRect(
                      gap: 2,
                      strokeWidth: 6,
                      color: Colors.red,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  Container(
                    height: 100,
                    width: 100,
                    color: Colors.yellow,
                    child: const DashedRect(
                      gap: 12,
                      strokeWidth: 1,
                    ),
                  ),
                  const SizedBox(
                    height: 100,
                    width: 100,
                    child: DashedRect(
                      gap: 2,
                      strokeWidth: 12,
                      color: Colors.red,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: const [
                  SizedBox(
                    height: 100,
                    width: 100,
                    child: DashedRect(
                      gap: 12,
                      strokeWidth: 12,
                    ),
                  ),
                  SizedBox(
                    height: 100,
                    width: 100,
                    child: DashedRect(
                      gap: 12,
                      strokeWidth: 1,
                      color: Colors.red,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: const [
                  SizedBox(
                    height: 100,
                    width: 100,
                    child: DashedRect(
                      gap: 10,
                      strokeWidth: 10,
                      color: Colors.orange,
                    ),
                  ),
                  SizedBox(
                    height: 100,
                    width: 100,
                    child: DashedRect(),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  Container(
                    height: 100,
                    width: 100,
                    decoration: BoxDecoration(
                      color: Colors.blue,
                      borderRadius: BorderRadius.circular(12.0),
                    ),
                    child: const DashedRect(
                      gap: 8,
                      strokeWidth: 8.0,
                      color: Colors.white,
                    ),
                  ),
                  Container(
                    height: 100,
                    width: 100,
                    decoration: BoxDecoration(
                      color: Colors.indigo,
                      borderRadius: BorderRadius.circular(12.0),
                    ),
                    child: const DashedRect(
                      gap: 8,
                      strokeWidth: 8.0,
                      color: Colors.orange,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 16.0),
            ],
          ),
        ),
      ),
    );
  }
}
